src/core/error_handling.ts

```diff
@@ -1,6 +1,101 @@
 import ErrorStackParser from "error-stack-parser";
 import { Module, API, Hiwire } from "./module.js";
 
+/**
+ * Dump the Python traceback to the browser console.
+ *
+ * @private
+ */
+API.dump_traceback = function () {
+  const fd_stdout = 1;
+  Module.__Py_DumpTraceback(fd_stdout, Module._PyGILState_GetThisThreadState());
+};
+
+let fatal_error_occurred = false;
+/**
+ * Signal a fatal error.
+ *
+ * Dumps the Python traceback, shows a JavaScript traceback, and prints a clear
+ * message indicating a fatal error. It then dummies out the public API so that
+ * further attempts to use Pyodide will clearly indicate that Pyodide has failed
+ * and can no longer be used. pyodide._module is left accessible, and it is
+ * possible to continue using Pyodide for debugging purposes if desired.
+ *
+ * @argument e {Error} The cause of the fatal error.
+ * @private
+ */
+API.fatal_error = function (e: any) {
+  if (e.pyodide_fatal_error) {
+    return;
+  }
+  if (fatal_error_occurred) {
+    console.error("Recursive call to fatal_error. Inner error was:");
+    console.error(e);
+    return;
+  }
+  if (typeof e === "number") {
+    // A C++ exception. Have to do some conversion work.
+    e = convertCppException(e);
+  }
+  // Mark e so we know not to handle it later in EM_JS wrappers
+  e.pyodide_fatal_error = true;
+  fatal_error_occurred = true;
+  console.error(
+    "Pyodide has suffered a fatal error. Please report this to the Pyodide maintainers."
+  );
+  console.error("The cause of the fatal error was:");
+  if (API.inTestHoist) {
+    // Test hoist won't print the error object in a useful way so convert it to
+    // string.
+    console.error(e.toString());
+    console.error(e.stack);
+  } else {
+    console.error(e);
+  }
+  try {
+    API.dump_traceback();
+    for (let key of Object.keys(API.public_api)) {
+      if (key.startsWith("_") || key === "version") {
+        continue;
+      }
+      Object.defineProperty(API.public_api, key, {
+        enumerable: true,
+        configurable: true,
+        get: () => {
+          throw new Error(
+            "Pyodide already fatally failed and can no longer be used."
+          );
+        },
+      });
+    }
+    if (API.on_fatal) {
+      API.on_fatal(e);
+    }
+  } catch (err2) {
+    console.error("Another error occurred while handling the fatal error:");
+    console.error(err2);
+  }
+  throw e;
+};
+
+class CppException extends Error {}
+Object.defineProperty(CppException.prototype, "name", {
+  value: CppException.name,
+});
+
+function convertCppException(ptr: number): CppException {
+  let msg;
+  try {
+    const msgPtr = Module._exc_what(ptr);
+    msg = Module.UTF8ToString(msgPtr);
+  } catch (e) {
```

Comment:
It might be a dumb question, but how can we catch this?
It found that [std::exception::what has a noexcept specifier](https://en.cppreference.com/w/cpp/error/exception/what).
Does Emscripten ignores noexcept specifier, or are there other reasons?

Reply:
We're catching a segfault...

Reply:
Oh... I got it. when it doesn't point to an exception it doesn't have `what` 🤦‍♀️